docs/root/intro/arch_overview/advanced/dynamic_modules.rst

```diff
@@ -0,0 +1,69 @@
+2. **Ease of Use**: The SDK should provide a high-level API that abstracts the details of the Envoy internals.
+3. **Flexibility**: The dynamic modules should be able to implement any functionality that can be implemented by the built-in C++ extensions without performance penalty. This is work in progress and many features are not yet available.
+
+Compatibility
+--------------------------
+
+Since the dynamic modules are loaded at runtime, it is important to ensure that the dynamic module is compatible with the
+Envoy binary that loads it.
+
+The dynamic modules have stricker compatibility requirements than the other Envoy extension mechanisms, such as Lua, Wasm or External Processor.
+This is partly because the ABI is tightly coupled with the Envoy internals that makes it a bit difficult to make a stable ABI. Even though
+our ultimate goal is to have a stable ABI that can be used across different versions of Envoy, we currently do not guarantee any compatibility
+between different versions.
+
+In other words, the dynamic modules must be built with the SDK of the same version as the Envoy binary that loads the dynamic module.
+Since the SDK lives inside the Envoy repository, using the same commit hash or release tag of the Envoy version is the best way to ensure
+the compatibility.
+
+Module Discovery
+--------------------------
+
+A dynamic module is referenced by its name as in the :ref:`configuration API  <envoy_v3_api_msg_extensions.dynamic_modules.v3.DynamicModuleConfig>`.
+The name is used to search for the shared library file in the search path. The search path is configured by the environment variable
+``ENVOY_DYNAMIC_MODULES_SEARCH_PATH``. The actual search path is ``${ENVOY_DYNAMIC_MODULES_SEARCH_PATH}/lib${name}.so``.
+
+For example, when the name ``my_module`` is referenced in the configuration and the search path is set to ``/path/to/modules``, Envoy will look for
+``/path/to/modules/libmy_module.so``.
+
+Safety
+--------------------------
+The dynamic modules should be used under the assumption that all modules are fully trusted and have the same privilege level as the main Envoy program.
+Since these modules run in the same process as Envoy, they can access all memory and resources available to the main process.
+This makes it unfeasible to enforce security boundaries between Envoy and the modules, as they share the same address space and permissions.
+It is essential that any dynamic module undergo thorough testing and validation before deployment just like any other application code.
+
+Getting Started
+--------------------------
+
+We have a dedicated repository for the dynamic module examples to help you get started.
+The repository is available at `envoyproxy/dynamic-modules-example <https://github.com/envoyproxy/envoyproxy/dynamic-modules-example>`_
```

Comment:
note: I chatted with @phlax offline that using the existing examples repository is not suitable for dynamic modules due to multiple reasons. One is because of the strict version compatibility requirement makes it impossible to make it part of the CI and the other is we want to show users that they don't need the envoy specific build settings like bazel etc to build modules and instead they can simply use the standard language tooling (at this point cargo/rustc). I will work with @phlax to set it up soon.